### ISSUE TITLE
Refresh access token if its about to expire 

### ISSUE THREAD

**Issue by kmarrip** [chaithanyaMarripati/goSpotify#15]
For every request access token ,refresh token are sent as cookies to the server 
- send an expiry time cookie which has unix time stamp of the expiry date, generate this when access token is generated 
- for every future request in the base handler, check the current time stamp and the expiry time stamp
- if it is about to expire in about lets say 5 min, send the request to refresh endpoint 
- get the refresh token from the cookie and talk to spotify via their api and save the access, refresh token and expiry unix time stamp 
- finally send the request to base url

**Comment by kmarrip** [comment]
@itmecho @StaniPetrosyan feel free to check out this issue

**Comment by itslychee** [comment]
I don't think it's really necessary to refresh the token 5 minutes in advance, why not just bump the expiration date of the `RefreshToken` cookie and renew the token when the `Token` cookie expires?

**Comment by kmarrip** [comment]
same thing either way

**Comment by itslychee** [comment]
> 
> 
> same thing either way

I agree, but by doing it at an earlier time, you are still going through the same process as if you were doing it when the token is gone. So would it not make sense to just not?

**Comment by kmarrip** [comment]
no, the thing is i dont want to rely on spotify to know if a token is expired. It would rather do it myself

**Comment by itslychee** [comment]
> 
> 
> no, the thing is i dont want to rely on spotify to know if a token is expired. It would rather do it myself

Well if the token expired how would you not rely on Spotify?

**Comment by kmarrip** [comment]
ok two things, if a token expires we need to refresh it. The question is when and how to do it. 
first problem --> how to know if a token is expired.( if it were a jwt, exp would be default present in it. It is not now).
                           either send a exp cookie back to browser or add expiry as token life time 
 what i wanted to do was send a exp cookie back and check if its about to expire in next 5 min. if it is then take the refresh token and send token cookies in response headers.


there could be other ways like check if token is missing but not refresh token. then we would simply refresh at that point. Though this may be a better way, i prefer doing it with exp cookie

**Comment by itslychee** [comment]
Couldn't you just rely on the browser since you're already storing the credentials in it? Why not just rely off of the expiration
of the cookies and check if they're gone, and if they are then why not just check if the RefreshToken cookie is alive? If it is then just utilize that cookie and ask spotify for a new token.

EDIT: My apologies, but it seems like you can't check the expiration date of the cookie
